### PR TITLE
feat: server-side heartbeat support

### PR DESCRIPTION
**Pull request by DMwangnima**
#### Check the PR title.

- [ ] This PR title match the format: \<type\>(optional scope): \<description\>
- [ ] The description of this PR title is user-oriented and clear enough for others to understand.
- [ ] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)

#### (Optional) More detailed description for this PR(en: English/zh: Chinese).

en:
zh(optional):

### CODE REVIEW

File: pkg/codec.go [hunk 1]
```diff
@@ -278,7 +282,7 @@ func (m *Hessian2Codec) Decode(ctx context.Context, message remote.Message, in r
 func (m *Hessian2Codec) decodeHeartbeatBody(ctx context.Context, header *dubbo.DubboHeader, message remote.Message, in remote.ByteBuffer) error {
 	// for heartbeat, there is no need to decode the body
 
-	// todo(DMwangnima): process heartbeat with UnknownHandler
+	message.SetMessageType(remote.Heartbeat)
```
Comment (felix021) [inline pkg/codec.go]: Do we need to consume the last Null?

Comment (DMwangnima) [reply pkg/codec.go]: dubbo-java use
```java
// in hessian2, this means the body is 'N'
Arrays.equals(payload, getNullBytesOf(getSerializationById(proto)))
```
to determine if this pkg is a heartbeat when IsEvent flag in dubbo header has been set.
As a result, we have to consume the entire body.

File: pkg/codec.go [hunk 2]
```diff
@@ -186,16 +190,15 @@ func (m *Hessian2Codec) encodeExceptionPayload(ctx context.Context, message remo
...
-//	// nil does not mean body is empty. after encoding, body contains 'N'
-//	if err := encoder.Encode(nil); err != nil {
-//		return nil, err
-//	}
-//
-//	return encoder.Buffer(), nil
-//}
+func (m *Hessian2Codec) encodeHeartbeatPayload(ctx context.Context, message remote.Message) (buf []byte, err error) {
+	encoder := hessian.NewEncoder()
+	// nil does not mean body is empty. after encoding, body contains 'N'
```
Comment (felix021) [inline pkg/codec.go]: This is not true now?

Comment (DMwangnima) [reply pkg/codec.go]: Event Flag set in dubbo header and 'N' body determines that this pkg is heartbeat.
For dubbo-go, it does not decode the body of the pkg when Event Flag is set in dubbo header.

For dubbo-java, it reads the body of the pkg and use this statement to judge when Event Flag is set in dubbo header.
```java
// in hessian2, this means the body is 'N'
Arrays.equals(payload, getNullBytesOf(getSerializationById(proto)))
```
For hessian2, NullByte is 'N'.

As a result, we need to encode nil in heartbeat response body for both dubbo-go side and dubbo-java side.

File: pkg/codec.go [hunk 3]
```diff
@@ -268,17 +271,25 @@ func (m *Hessian2Codec) Decode(ctx context.Context, message remote.Message, in r
...
+	body, err := readBody(header, in)
+	if err != nil {
+		return err
+	}
+
+	// entire body equals to BC_NULL determines that this request is a heartbeat
+	if len(body) == 1 && body[0] == commons.BC_NULL {
+		message.SetMessageType(remote.Heartbeat)
+	}
+	// there are other events(READONLY_EVENT, WRITABLE_EVENT) in dubbo-java that we are not planing to implement
```
Comment (felix021) [inline pkg/codec.go]: plan`n`ing to implement `currently`.

File: pkg/codec.go (hunk 2 above)
Comment (felix021) [reply pkg/codec.go]: Oh I see. I took it wrong as the N appended.

**Review by felix021 — APPROVED** [review]

**Pull request merged**